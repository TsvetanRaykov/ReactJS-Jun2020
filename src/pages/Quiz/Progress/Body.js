import React, { useState, useEffect, Fragment } from 'react'

import {
	Paper,
	Box,
	RadioGroup,
	Radio,
	Typography,
	FormControlLabel,
	Container,
	FormControl,
	FormLabel,
} from '@material-ui/core'

const QuizActiveBody = (props) => {
	const { handleAnswer, question } = props
	const [answers, setAnswers] = useState([])

	const [value, setValue] = useState(question.userAnswer || '')
	const handleChange = (event) => {
		setValue(event.target.value)
		handleAnswer(event.target.value)
	}

	useEffect(() => {
		setAnswers(question?.answers || [])
		setValue(question.userAnswer || '')
	}, [question])

	return (
		<Fragment>
			<Container>
				<Paper>
					<Box m={2} px={3} py={2} display='flex'>
						<Typography variant='h6'>{question?.question}</Typography>
					</Box>
				</Paper>
			</Container>
			<Container>
				<Paper>
					<Box
						display='flex'
						flexDirection='column'
						alignItems='flex-start'
						px={5}
						py={2}
					>
						<FormControl component='fieldset'>
							<FormLabel component='legend'></FormLabel>
							<RadioGroup value={value} onChange={handleChange}>
								{answers.map((a) => (
									<FormControlLabel
										key={a.text}
										value={a.text}
										control={<Radio color='primary' />}
										label={a.text}
									/>
								))}
							</RadioGroup>
						</FormControl>
					</Box>
				</Paper>
			</Container>
		</Fragment>
	)
}

export default QuizActiveBody
