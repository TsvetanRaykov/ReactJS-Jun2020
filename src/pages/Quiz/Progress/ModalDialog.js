import React, { useState } from 'react'
import { makeStyles } from '@material-ui/core/styles'
import Modal from '@material-ui/core/Modal'
import Backdrop from '@material-ui/core/Backdrop'
import Fade from '@material-ui/core/Fade'

const useStyles = makeStyles((theme) => ({
	modal: {
		display: 'flex',
		alignItems: 'center',
		justifyContent: 'center',
	},
	paper: {
		backgroundColor: theme.palette.background.paper,
		border: '2px solid #000',
		boxShadow: theme.shadows[5],
		padding: theme.spacing(2, 4, 3),
	},
}))

export default function ModalDialog() {
	const classes = useStyles()
	const [open, setOpen] = useState(false)

	const handleOpen = () => {
		setOpen(true)
	}

	const handleClose = () => {
		setOpen(false)
	}

	return (
		<div>
			<button type='button' onClick={handleOpen}>
				modal-dialog
			</button>
			<Modal
				aria-labelledby='modal-dialog-title'
				aria-describedby='modal-dialog-description'
				className={classes.modal}
				open={open}
				onClose={handleClose}
				closeAfterTransition
				BackdropComponent={Backdrop}
				BackdropProps={{
					timeout: 500,
				}}
			>
				<Fade in={open}>
					<div className={classes.paper}>
						<h2 id='modal-dialog-title'>Title</h2>
						<p id='modal-dialog-description'>text</p>
					</div>
				</Fade>
			</Modal>
		</div>
	)
}
