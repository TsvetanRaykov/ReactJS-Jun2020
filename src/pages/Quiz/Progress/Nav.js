import React from 'react'

import {
	Paper,
	Box,
	Button,
	ButtonGroup,
	createTheme,
	ThemeProvider,
	Container,
	Grid,
} from '@material-ui/core'
import { green } from '@material-ui/core/colors'

const QuizNav = ({ questions, changeQuestion, handleEndQuizClick }) => {
	const handleClick = (e, i) => {
		changeQuestion(i)
	}
	const theme = createTheme({
		palette: {
			primary: {
				main: green[500],
				contrastText: '#fff',
			},
		},
	})

	return (
		<ThemeProvider theme={theme}>
			<Container>
				<Paper>
					<Grid container>
						<Grid item xs={10}>
							<Box
								display='flex'
								justifyContent='space-around'
								flexDirection='row'
								p={3}
							>
								<ButtonGroup
									disableElevation
									variant='outlined'
									color='primary'
								>
									{questions.map((n, i) => (
										<Button
											key={i}
											variant={
												n.userAnswers && n.userAnswers.length > 0
													? 'contained'
													: 'outlined'
											}
											onClick={(e) => handleClick(e, i)}
										>
											{i + 1}
										</Button>
									))}
								</ButtonGroup>
							</Box>
						</Grid>
						<Grid item xs={2}>
							<Box
								display='flex'
								justifyContent='space-around'
								flexDirection='row'
								p={3}
							>
								<Button
									variant='contained'
									color='primary'
									onClick={handleEndQuizClick}
								>
									End
								</Button>
							</Box>
						</Grid>
					</Grid>
				</Paper>
			</Container>
		</ThemeProvider>
	)
}

export default QuizNav
