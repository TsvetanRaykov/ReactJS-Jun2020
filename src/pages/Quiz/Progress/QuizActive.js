import React, { Fragment, useState } from 'react'

import QuizNav from './Nav'
import QuizActiveBody from './Body'

const QuizActive = (props) => {
	const {
		quiz: {
			data: { questions },
		},
		handleEndQuizClick,
	} = props

	const [active, setActiveQuestion] = useState(0)

	const handleAnswer = (a, remove) => {
		const question = questions[active]

		if (!question?.userAnswers) {
			question.userAnswers = []
		}

		if (question.type === 'multiple') {
			const set = new Set(question.userAnswers)
			remove ? set.delete(a) : set.add(a)
			question.userAnswers = Array.from(set)
		} else {
			if (a.trim() === '') question.userAnswers = []
			else question.userAnswers = [a]
		}
		console.log(question.userAnswers)
	}

	const changeQuestion = (i) => {
		setActiveQuestion(i)
	}

	return (
		<Fragment>
			<QuizNav
				questions={questions}
				changeQuestion={changeQuestion}
				handleEndQuizClick={handleEndQuizClick}
			/>
			<QuizActiveBody
				question={questions[active]}
				handleAnswer={handleAnswer}
			/>
		</Fragment>
	)
}

export default QuizActive
