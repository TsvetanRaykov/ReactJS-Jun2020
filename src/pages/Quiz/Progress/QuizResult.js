import React from 'react'
import { Paper, Box, Typography } from '@material-ui/core'
import { padZero } from '../../../utils'

const QuizRezult = (props) => {
	const { correct, total, duration } = props
	const minutes = padZero(parseInt(duration / 60, 10))
	const seconds = padZero(parseInt(duration % 60, 10))
	return (
		<Paper>
			<Box
				display='flex'
				my={3}
				py={2}
				flexDirection='column'
				justifyContent='center'
			>
				<Box display='flex' py={2} flexDirection='row' justifyContent='center'>
					<h5>Your score is</h5>
				</Box>

				<Box display='flex' py={2} flexDirection='row' justifyContent='center'>
					<h1>{`${correct} / ${total}`}</h1>
				</Box>
			</Box>
			<Box display='flex' pb={2} flexDirection='row' justifyContent='center'>
				<Typography>{`You pass it in: ${minutes} : ${seconds}`}</Typography>
			</Box>
		</Paper>
	)
}

export default QuizRezult
