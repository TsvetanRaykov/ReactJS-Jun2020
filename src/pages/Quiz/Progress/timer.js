import React, { Fragment, useState, useEffect } from 'react'

const QuizTimer = (props) => {
	const { timer } = props
	const { duration, start, complete } = timer || {}
	const [minutes, setMinutes] = useState(0)
	const [seconds, setSeconds] = useState(0)

	const [timerHandler, setTimerHandler] = useState(0)

	const startTimer = () => {
		const interval = 1000 // ms
		let timer = duration
		let expected = Date.now() + interval
		setTimerHandler(() => {
			return setTimeout(step, interval)
		})

		function step() {
			const dt = Date.now() - expected

			setMinutes(parseInt(timer / 60, 10))
			setSeconds(parseInt(timer % 60, 10))

			if (--timer < 0) {
				complete()
				clearTimeout(timerHandler)
				return
			}

			expected += interval
			setTimerHandler(() => {
				return setTimeout(step, Math.max(0, interval - dt))
			})
			console.log(timerHandler)
		}
	}

	const padZero = (num) => (num < 10 ? `0${num}` : `${num}`)

	useEffect(() => {
		if (start) {
			startTimer()
		}

		return () => {
			clearTimeout(timerHandler)
		}
		// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [start])

	return <Fragment>{`${padZero(minutes)} : ${padZero(seconds)}`}</Fragment>
}

export default QuizTimer
