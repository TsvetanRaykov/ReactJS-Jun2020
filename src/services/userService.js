import app from './firebase'
import 'firebase/auth'
import 'firebase/firebase-firestore'
import 'firebase/storage'

class Firebase {
	constructor() {
		this.auth = app.auth()
		this.db = app.firestore()
		this.storage = app.storage()
	}

	login(email, password) {
		return this.auth.signInWithEmailAndPassword(email, password)
	}

	async logout() {
		await this.auth.signOut()
	}

	async register(name, email, password) {
		await this.auth.createUserWithEmailAndPassword(email, password)
		await this.auth.currentUser.updateProfile({
			displayName: name,
		})
	}

	updateName(newName) {
		this.auth.currentUser.updateProfile({
			displayName: newName,
		})
	}

	addImage(imageFile) {
		if (!this.auth.currentUser) {
			return alert('Not authorized')
		}
		const uploadTask = this.storage
			.ref(`/images/${imageFile.name}`)
			.put(imageFile)

		uploadTask.on(
			'state_changed',
			(snapShot) => {
				console.log(snapShot)
			},
			(err) => {
				console.error(err)
			},
			async () => {
				const photoURL = await this.storage
					.ref('images')
					.child(imageFile.name)
					.getDownloadURL()

				await this.auth.currentUser.updateProfile({
					photoURL,
				})
			}
		)
	}

	getCurrentUser() {
		const user = this.auth.currentUser
			? {
					userName: this.auth.currentUser.displayName,
					userImg: this.auth.currentUser.photoURL || '-',
					userEmail: this.auth.currentUser.email,
					userId: this.auth.currentUser.uid,
			  }
			: null
		return user
	}

	isInitialized() {
		return new Promise((resolve) => {
			this.auth.onAuthStateChanged(resolve)
		})
	}
}

export default new Firebase()
