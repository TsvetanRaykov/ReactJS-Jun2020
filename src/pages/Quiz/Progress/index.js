import React, { Fragment, useState, useEffect } from 'react'
import { withRouter } from 'react-router-dom'
import Header from '../../../components/Header'
import {
	Button,
	Paper,
	Typography,
	Box,
	ThemeProvider,
	createMuiTheme,
	Container,
	makeStyles,
} from '@material-ui/core'
import Loader from '../../../components/Loader'
import { green } from '@material-ui/core/colors'
import QuizActive from './QuizActive'
import ModalDialog from '../../../components/shared/ModalDialog'
import QuizRezult from './QuizResult'
import { processQuizResult } from '../../../utils'

import quizService from '../../../services/quizService'

const useStyles = makeStyles((theme) => ({
	button: {
		margin: theme.spacing(1),
	},
}))
const QuizProgress = (props) => {
	const classes = useStyles()
	const {
		match: {
			params: { id },
		},
		history,
	} = props

	const [quiz, setQuiz] = useState({})
	const [loading, setLoading] = useState(false)
	const [start, setStart] = useState(false)
	const [timeLeft, setTimeLeft] = useState(0)
	const [quizResult, setQuizResult] = useState(null)

	const [modalDialog, setModalDialog] = useState({
		title: '',
		message: '',
		open: false,
		handleYes: () => {},
		handleNo: () => {},
	})

	useEffect(() => {
		setLoading(true)
		quizService
			.getById(atob(id))
			.then((data) => {
				setQuiz(data)
				setTimer((t) => {
					return { ...t, duration: data.data.duration }
				})
			})
			.finally(() => setLoading(false))
	}, [id])

	const finishQuiz = () => {
		setStart(false)
		setTimer(() => {
			setQuizResult(() => {
				const result = processQuizResult(quiz)
				quizService.completeQuiz({
					quiz,
					correct: result[0],
					total: result[1],
					timeLeft,
				})
				return result
			})
			return {
				start: false,
				duration: timeLeft,
			}
		})
	}

	const buildEndQuizConfirmationDialog = () => ({
		title: 'Are you ready to end the quiz?',
		message: 'You will not be able to continue.',
		open: true,
		handleYes: () => {
			setModalDialog({ open: false })
			finishQuiz()
		},
		handleNo: () => {
			setModalDialog({ open: false })
		},
	})

	history.block(function () {
		if (start) {
			setModalDialog(buildEndQuizConfirmationDialog())
			return false
		}
	})

	const handleEndQuizClick = () => {
		setModalDialog(buildEndQuizConfirmationDialog())
	}

	const [timer, setTimer] = useState({
		duration: 0,
		start: false,
		complete: () => {},
		progress: () => {},
	})

	const handleQuizStartClick = () => {
		setModalDialog({
			title: 'Are you ready to start?',
			message: '',
			open: true,
			handleYes: () => {
				setModalDialog({ open: false })
				setStart(true)

				setTimer(() => ({
					duration: quiz?.data.duration,
					start: true,
					complete: () => {
						finishQuiz()
					},
					progress: (s) => {
						setTimeLeft(s)
					},
				}))
			},
			handleNo: () => {
				setModalDialog({ open: false })
			},
		})
	}

	const theme = createMuiTheme({
		palette: {
			primary: {
				main: green[500],
				contrastText: '#fff',
			},
		},
	})

	const renderQuizResult = () => {
		if (!quizResult) {
			return
		}
		const [correct, total] = quizResult

		return (
			<QuizRezult
				correct={correct}
				total={total}
				duration={quiz?.data.duration - timer.duration}
			/>
		)
	}

	const renderButtonStart = () => (
		<ThemeProvider theme={theme}>
			<Box display='flex' py={2} flexDirection='row' justifyContent='center'>
				<Button
					className={classes.button}
					variant='contained'
					color='secondary'
					onClick={() => {
						history.goBack()
					}}
				>
					Back
				</Button>
				<Button
					className={classes.button}
					variant='contained'
					color='primary'
					onClick={handleQuizStartClick}
				>
					Start
				</Button>
			</Box>
		</ThemeProvider>
	)

	const renderQuizIntro = () => {
		const { data } = quiz

		if (start) {
			return <QuizActive quiz={quiz} handleEndQuizClick={handleEndQuizClick} />
		}

		return (
			<Container>
				<Paper>
					<Box
						display='flex'
						my={3}
						py={2}
						flexDirection='row'
						justifyContent='center'
					>
						<Typography component='h1' variant='h4'>
							{data?.title}
						</Typography>
					</Box>
					<Box
						display='flex'
						py={2}
						flexDirection='row'
						justifyContent='center'
					>
						<Typography>{data?.description}</Typography>
					</Box>
				</Paper>
				{quizResult ? renderQuizResult() : renderButtonStart()}
			</Container>
		)
	}

	return (
		<Fragment>
			<Header timer={timer} />
			<ModalDialog {...modalDialog} />
			{loading ? <Loader /> : renderQuizIntro()}
		</Fragment>
	)
}

export default withRouter(QuizProgress)
