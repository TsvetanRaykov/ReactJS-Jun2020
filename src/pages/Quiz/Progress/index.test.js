import React from 'react'
import { act, create } from 'react-test-renderer'
import TestWrapper from '../../../utils/test-wrapper'
import Progress from '.'
import Body from './Body'
import Nav from './Nav'
import QuizActive from './QuizActive'
import QuizResult from './QuizResult'
import userService from '../../../services/userService'

import { shallow } from 'enzyme'

jest.mock('react', () => ({
	...jest.requireActual('react'),
	useLayoutEffect: jest.requireActual('react').useEffect,
}))

describe('<Progress />', () => {
	it('Should match the snapshot', () => {
		jest.spyOn(React, 'useEffect').mockImplementation((f) => f())
		jest.spyOn(userService, 'getCurrentUser').mockImplementation(() => {
			return {}
		})

		act(() => {
			const match = { params: { id: 1 }, isExact: true, path: '', url: '' }

			const wrapper = shallow(
				<TestWrapper>
					<Progress match={match} />
				</TestWrapper>
			)
			expect(wrapper).toMatchSnapshot()
		})
	})
})

describe('<Body />', () => {
	it('Should match the snapshot', () => {
		act(() => {
			const wrapper = create(
				<TestWrapper>
					<Body question={{}} />
				</TestWrapper>
			).toJSON()
			expect(wrapper).toMatchSnapshot()
		})
	})
})

describe('<Nav />', () => {
	it('Should match the snapshot', () => {
		act(() => {
			const wrapper = create(
				<TestWrapper>
					<Nav questions={[]} />
				</TestWrapper>
			).toJSON()
			expect(wrapper).toMatchSnapshot()
		})
	})
})

describe('<QuizActive />', () => {
	it('Should match the snapshot', () => {
		act(() => {
			const wrapper = create(
				<TestWrapper>
					<QuizActive quiz={{ data: { questions: [{ userAnswers: [] }] } }} />
				</TestWrapper>
			).toJSON()
			expect(wrapper).toMatchSnapshot()
		})
	})
})

describe('<QuizResult />', () => {
	it('Should match the snapshot', () => {
		act(() => {
			const wrapper = create(
				<TestWrapper>
					<QuizResult />
				</TestWrapper>
			).toJSON()
			expect(wrapper).toMatchSnapshot()
		})
	})
})
