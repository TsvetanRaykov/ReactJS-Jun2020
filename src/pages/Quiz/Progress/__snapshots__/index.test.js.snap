// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Body /> Should match the snapshot 1`] = `null`;

exports[`<Nav /> Should match the snapshot 1`] = `null`;

exports[`<Progress /> Should match the snapshot 1`] = `ShallowWrapper {}`;

exports[`<QuizActive /> Should match the snapshot 1`] = `null`;

exports[`<QuizResult /> Should match the snapshot 1`] = `null`;
